In [1]:
import pandas as pd
import numpy as np


In [2]:
df = pd.read_excel(r"Telco_customer_churn.xlsx")

In [4]:
print("Shape of dataset: ",df.shape)
df.info()

Shape of dataset:  (7043, 33)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   object 
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   object 
 3   State              7043 non-null   object 
 4   City               7043 non-null   object 
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   object 
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   object 
 10  Senior Citizen     7043 non-null   object 
 11  Partner            7043 non-null   object 
 12  Dependents         7043 non-null   object 
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   object 
 15  Multiple Lines     7043 non-null   object 

In [33]:
#Check missing values
missing_values=df.isnull().sum()
print("Missing values : ",missing_values)

Missing values :  CustomerID              0
Count                   0
Country                 0
State                   0
City                    0
Zip Code                0
Lat Long                0
Latitude                0
Longitude               0
Gender                  0
Senior Citizen          0
Partner                 0
Dependents              0
Tenure Months           0
Phone Service           0
Multiple Lines          0
Internet Service        0
Online Security         0
Online Backup           0
Device Protection       0
Tech Support            0
Streaming TV            0
Streaming Movies        0
Contract                0
Paperless Billing       0
Payment Method          0
Monthly Charges         0
Total Charges           0
Churn Label             0
Churn Value             0
Churn Score             0
CLTV                    0
Churn Reason         5174
dtype: int64


Churn Reason has 5174 missing values . Only 1869 values are present

In [39]:
Duplicate_IDs=df['CustomerID'].duplicated().sum()
print("Duplicate customer id :",Duplicate_IDs)

Duplicate customer id : 0


No duplicates

In [40]:
#Check invalid monhtly charges
invalid_monthly_charge=df[df['Monthly Charges'] < 0]
print("Invalid monthly charge",len(invalid_monthly_charge))

Invalid monthly charge 0


In [14]:
df['Monthly Charges'] == 0

0       False
1       False
2       False
3       False
4       False
        ...  
7038    False
7039    False
7040    False
7041    False
7042    False
Name: Monthly Charges, Length: 7043, dtype: bool

No invalid monthly charges

In [16]:
#Check Invalid tenure months
Invalid_tenure_months=df[df['Tenure Months'] < 0]
print("Invalid tenure months ",len(Invalid_tenure_months))

Invalid tenure months  0


No invalid tenure months

In [ ]:
#Check invalid cltv(Customer Lifetime Value)
invalid_cltv = df[df['CLTV'] < 0]

print("Invalid CLTV:", len(invalid_cltv))



Invalid CLTV: 0


No invalid cltv

In [23]:
#Check if blank data in total charges column
blank_charges = df[df['Total Charges'].isna()]
print("Rows with blank Total Charges:",blank_charges)


Rows with blank Total Charges: Empty DataFrame
Columns: [CustomerID, Count, Country, State, City, Zip Code, Lat Long, Latitude, Longitude, Gender, Senior Citizen, Partner, Dependents, Tenure Months, Phone Service, Multiple Lines, Internet Service, Online Security, Online Backup, Device Protection, Tech Support, Streaming TV, Streaming Movies, Contract, Paperless Billing, Payment Method, Monthly Charges, Total Charges, Churn Label, Churn Value, Churn Score, CLTV, Churn Reason]
Index: []

[0 rows x 33 columns]


In [31]:
categorical_cols = df.select_dtypes(include='object').columns

for col in categorical_cols:
    print(f"\n{col}")
    print(df[col].unique())


CustomerID
['3668-QPYBK' '9237-HQITU' '9305-CDSKC' ... '2234-XADUH' '4801-JZAZL'
 '3186-AJIEK']

Country
['United States']

State
['California']

City
['Los Angeles' 'Beverly Hills' 'Huntington Park' ... 'Standish' 'Tulelake'
 'Olympic Valley']

Lat Long
['33.964131, -118.272783' '34.059281, -118.30742' '34.048013, -118.293953'
 ... '40.346634, -120.386422' '41.813521, -121.492666'
 '39.191797, -120.212401']

Gender
['Male' 'Female']

Senior Citizen
['No' 'Yes']

Partner
['No' 'Yes']

Dependents
['No' 'Yes']

Phone Service
['Yes' 'No']

Multiple Lines
['No' 'Yes' 'No phone service']

Internet Service
['DSL' 'Fiber optic' 'No']

Online Security
['Yes' 'No' 'No internet service']

Online Backup
['Yes' 'No' 'No internet service']

Device Protection
['No' 'Yes' 'No internet service']

Tech Support
['No' 'Yes' 'No internet service']

Streaming TV
['No' 'Yes' 'No internet service']

Streaming Movies
['No' 'Yes' 'No internet service']

Contract
['Month-to-month' 'Two year' 'One year']

Paper

In [42]:
quality_report = pd.DataFrame({
    'Issue': [
        'missing_values',
        'Duplicate_IDs',
        'invalid_monthly_charge',
        'Invalid_tenure_months',
        'invalid_cltv',
        'blank_charges'
    ],
    'Count': [
        df.isnull().sum().sum(),
        df['CustomerID'].duplicated().sum(),
        len(invalid_monthly_charge),
        len(Invalid_tenure_months),
        len(invalid_cltv),
        len(blank_charges)
    ]
})

quality_report

,Issue,Count
0,missing_values,5174
1,Duplicate_IDs,0
2,invalid_monthly_charge,0
3,Invalid_tenure_months,0
4,invalid_cltv,0
5,blank_charges,0


5174 missing values are usually from the Churn Reason column because customers who haven't churned don't have a churn reason.

In [43]:
quality_report.to_excel("Data_Quality_Report.xlsx")